tests/Unit/ActionTesting.hpp

```diff
@@ -61,6 +61,260 @@ struct get_array_index;
+ * the test snippet above. The component and tag are passed as template
+ * parameters while the runner and array index are passed as arguments. It is
+ * also possible to retrieve DataBox directly using the
+ * `ActionTesting::get_databox()` function if the DataBox type is known:
+ *
+ * \snippet Test_ActionTesting.cpp get databox
+ *
+ * It is also possible to check if an item can be retrieved from the DataBox
+ * using the `ActionTesting::tag_is_retrievable()` function as follows:
+ *
+ * \snippet Test_ActionTesting.cpp tag is retrievable
+ *
+ * ### Stub actions and invoking simple and threaded actions
+ *
+ * A simple action can be invoked on a distributed object or component using
+ * the `ActionTesting::simple_action()` function:
+ *
+ * \snippet Test_ActionTesting.cpp invoke simple action
+ *
+ * A threaded action can be invoked on a distributed object or component using
+ * the `ActionTesting::threaded_action()` function:
+ *
+ * \snippet Test_ActionTesting.cpp invoke threaded action
+ *
+ * Sometimes an individual action calls other actions but we still want to be
+ * able to test the action and its effects in isolation. To this end the ATF
+ * supports replacing calls to actions with calls to other actions. For example,
+ * instead of calling `simple_action_a` we want to call `simple_action_a_mock`
+ * which just verifies the data received is correct and sets a flag in the
+ * DataBox that it was invoked. This can be done by setting the type aliases
+ * `replace_these_simple_actions` and `with_these_simple_actions` in the
+ * parallel component definition as follows:
+ *
+ * \snippet Test_ActionTesting.cpp simple action replace
+ *
+ * The length of the two type lists must be the same and the Nth action in
+ * `replace_these_simple_actions` gets replaced by the Nth action in
+ * `with_these_simple_actions`. The same feature also exists for threaded
+ * actions:
+ *
+ * \snippet Test_ActionTesting.cpp threaded action replace
+ *
+ * Furthermore, simple actions invoked from an action are not run immediately.
+ * Instead, they are queued so that order randomization and introspection may
+ * occur. The simplest form of introspection is checking whether the simple
+ * action queue is empty:
+ *
+ * \snippet Test_ActionTesting.cpp simple action queue empty
+ *
+ * The `ActionTesting::invoke_queued_simple_action()` invokes the next queued
+ * simple action:
+ *
+ * \snippet Test_ActionTesting.cpp invoke queued simple action
+ *
+ * Note that the same functionality exists for threaded actions. The functions
+ * are `ActionTesting::is_threaded_action_queue_empty()`, and
+ * `ActionTesting::invoke_queued_threaded_action()`.
+ *
+ * ### Mocking or replacing components with stubs
+ *
+ * An action can invoke an action on another parallel component. In this case we
+ * need to be able to tell the mocking framework to replace the component the
+ * action is trying to call and instead call a different component that we have
```

Comment:
should this refer to calling actions on components rather than calling components?